source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -0,0 +1,36 @@
+#include "MapdataCheckPoint.hh"
+
+#include "game/system/CourseMap.hh"
+
+namespace System {
+
+/// @addr{0x805154E4}
+MapdataCheckPoint::MapdataCheckPoint(const SData *data)
+    : m_rawData(data), m_nextCount(0), m_prevCount(0) {
+    u8 *unsafeData = reinterpret_cast<u8 *>(const_cast<SData *>(data));
+    EGG::RamStream stream = EGG::RamStream(unsafeData, sizeof(SData));
+    read(stream);
+    m_midpoint = (m_left + m_right) / 2.0f;
```

Comment:
```suggestion
    m_midpoint = 0.5f * (m_left + m_right);
```
This is one of the most common "optimization" mistakes I see people make. Division and multiplication are *not* necessarily inverse operations with floating-point numbers, nor is multiplication commutative. Therefore, we need to examine the instructions carefully.

Reply:
i had to swap the order to `(m_left + m_right) * 0.5f` to compile. i assume thats because operator* isn't comutative by default?

Reply:
A friend operator needs to be added, then.